File: srcs/data_plat_digiedu/mixed_effects_analysis.py
"""
====================================================================
MIXED EFFECTS MODEL ANALYSIS - PYTHON VERSION
Teacher Perception of Student Performance Study
====================================================================

Required packages:
pip install pandas numpy statsmodels matplotlib seaborn scipy
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.regression.mixed_linear_model import MixedLM
from statsmodels.formula.api import mixedlm
import scipy.stats as stats
from datetime import datetime

# Set style
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

# ====================================================================
# 1. LOAD AND PREPARE DATA
# ====================================================================

print("=" * 70)
print("LOADING DATA")
print("=" * 70)

# Load data
df = pd.read_csv("data_structure_template.csv")

# Convert date
df['survey_date'] = pd.to_datetime(df['survey_date'])

# Calculate time variable (years from first survey)
min_date = df['survey_date'].min()
df['time_years'] = (df['survey_date'] - min_date).dt.days / 365.25

# Recode intervention (Treatment=1, Control=0)
df['intervention'] = (df['intervention_status'] == 'Treatment').astype(int)

# Center continuous predictors
df['teaching_exp_centered'] = df['teaching_experience_years'] - df['teaching_experience_years'].mean()
df['class_size_centered'] = df['class_size'] - df['class_size'].mean()

# Check data
print("\nData shape:", df.shape)
print("\nFirst few rows:")
print(df[['teacher_id', 'school_id', 'region', 'intervention', 'time_years', 
          'b1_overall_performance']].head())

# ====================================================================
# 2. DESCRIPTIVE STATISTICS
# ====================================================================

print("\n" + "=" * 70)
print("DESCRIPTIVE STATISTICS")
print("=" * 70)

print("\n--- Sample Sizes ---")
print(f"Total responses: {len(df)}")
print(f"Number of teachers: {df['teacher_id'].nunique()}")
print(f"Number of schools: {df['school_id'].nunique()}")
print(f"Number of regions: {df['region'].nunique()}")

print("\n--- By Intervention Status ---")
print(df['intervention_status'].value_counts())

print("\n--- Cross-tabulation: Region x Intervention ---")
print(pd.crosstab(df['region'], df['intervention_status']))

print("\n--- Means by Intervention Status ---")
summary_by_intervention = df.groupby('intervention_status').agg({
    'b1_overall_performance': ['count', 'mean', 'std'],
    'b2_problem_solving': ['mean', 'std'],
    'b6_engagement': ['mean', 'std']
}).round(3)
print(summary_by_intervention)

print("\n--- Means by Region ---")
summary_by_region = df.groupby('region').agg({
    'b1_overall_performance': ['count', 'mean', 'std']
}).round(3)
print(summary_by_region)

# ====================================================================
# 3. NULL MODEL (Variance Partitioning)
# ====================================================================

print("\n" + "=" * 70)
print("NULL MODEL: VARIANCE PARTITIONING")
print("=" * 70)

# Note: Python's statsmodels doesn't easily handle nested random effects
# We'll use school nested in region by creating a composite grouping variable

# Create a composite school identifier that accounts for nesting
df['school_in_region'] = df['region'].astype(str) + "_" + df['school_id'].astype(str)

# Fit null model (random intercept for schools only)
model0 = mixedlm("b1_overall_performance ~ 1", 
                 data=df, 
                 groups=df["school_id"])
result0 = model0.fit(reml=True)

print(result0.summary())

# Calculate ICC
var_school = float(result0.cov_re.iloc[0, 0])
var_residual = result0.scale
ICC_school = var_school / (var_school + var_residual)

print("\n--- Variance Components ---")
print(f"School variance: {var_school:.3f}")
print(f"Residual variance: {var_residual:.3f}")
print(f"\n--- Intraclass Correlation ---")
print(f"ICC (School): {ICC_school:.3f}")

if ICC_school > 0.05:
    print("\n✓ ICC > 0.05: Mixed effects model is justified!")
else:
    print("\n⚠ ICC < 0.05: Consider whether mixed effects model is necessary")

# ====================================================================
# 4. MODEL 1: MAIN EFFECTS
# ====================================================================

print("\n" + "=" * 70)
print("MODEL 1: MAIN EFFECTS OF INTERVENTION AND TIME")
print("=" * 70)

model1 = mixedlm("b1_overall_performance ~ intervention + time_years", 
                 data=df, 
                 groups=df["school_id"])
result1 = model1.fit(reml=False)

print(result1.summary())

# ====================================================================
# 5. MODEL 2: INTERVENTION × TIME INTERACTION
# ====================================================================

print("\n" + "=" * 70)
print("MODEL 2: INTERVENTION × TIME INTERACTION")
print("=" * 70)

model2 = mixedlm("b1_overall_performance ~ intervention * time_years", 
                 data=df, 
                 groups=df["school_id"])
result2 = model2.fit(reml=False)

print(result2.summary())

# Likelihood ratio test
lr_stat = -2 * (result1.llf - result2.llf)
df_diff = len(result2.params) - len(result1.params)
p_value = stats.chi2.sf(lr_stat, df_diff)

print("\n--- Model Comparison ---")
print(f"Likelihood Ratio Test: χ² = {lr_stat:.3f}, df = {df_diff}, p = {p_value:.4f}")

# ====================================================================
# 6. MODEL 3: ADD COVARIATES (FULL MODEL)
# ====================================================================

print("\n" + "=" * 70)
print("MODEL 3: FULL MODEL WITH COVARIATES")
print("=" * 70)

model3 = mixedlm("b1_overall_performance ~ intervention * time_years + "
                 "C(region) + teaching_exp_centered + class_size_centered", 
                 data=df, 
                 groups=df["school_id"])
result3 = model3.fit(reml=False)

print(result3.summary())

# Calculate marginal and conditional R²
# Marginal R² (fixed effects only)
y_pred_fixed = result3.fittedvalues
y_actual = df['b1_overall_performance'].dropna()
# Align indices
y_pred_fixed = y_pred_fixed[y_actual.index]
r2_marginal = 1 - np.var(y_actual - y_pred_fixed) / np.var(y_actual)

print(f"\n--- Model Fit ---")
print(f"Marginal R² (fixed effects): {r2_marginal:.3f}")

# ====================================================================
# 7. ANALYZE ALL OUTCOMES
# ====================================================================

print("\n" + "=" * 70)
print("MODELS FOR ALL OUTCOMES")
print("=" * 70)

outcomes = ['b1_overall_performance', 'b2_problem_solving', 
            'b3_critical_thinking', 'b4_collaboration',
            'b5_communication', 'b6_engagement', 
            'b7_behavior', 'b8_persistence']

results_dict = {}
results_summary = []

for outcome in outcomes:
    print(f"\n--- Outcome: {outcome} ---")
    
    formula = f"{outcome} ~ intervention * time_years + C(region) + teaching_exp_centered + class_size_centered"
    
    try:
        model = mixedlm(formula, data=df, groups=df["school_id"])
        result = model.fit(reml=False, method='powell')  # Use powell for better convergence
        results_dict[outcome] = result
        
        # Extract key coefficients
        params = result.params
        pvalues = result.pvalues
        
        results_summary.append({
            'Outcome': outcome,
            'Intercept': params.get('Intercept', np.nan),
            'Intervention_Effect': params.get('intervention', np.nan),
            'Intervention_p': pvalues.get('intervention', np.nan),
            'Time_Effect': params.get('time_years', np.nan),
            'Time_p': pvalues.get('time_years', np.nan),
            'Interaction_Effect': params.get('intervention:time_years', np.nan),
            'Interaction_p': pvalues.get('intervention:time_years', np.nan)
        })
        
        print(f"✓ Model converged successfully")
        print(f"  Intervention effect: {params.get('intervention', np.nan):.3f} (p = {pvalues.get('intervention', np.nan):.4f})")
        
    except Exception as e:
        print(f"✗ Model failed: {str(e)}")
        results_summary.append({
            'Outcome': outcome,
            'Intercept': np.nan,
            'Intervention_Effect': np.nan,
            'Intervention_p': np.nan,
            'Time_Effect': np.nan,
            'Time_p': np.nan,
            'Interaction_Effect': np.nan,
            'Interaction_p': np.nan
        })

# Create summary dataframe
results_df = pd.DataFrame(results_summary)
print("\n" + "=" * 70)
print("SUMMARY RESULTS TABLE")
print("=" * 70)
print(results_df.round(4))

# Save results
results_df.to_csv("mixed_effects_results_summary_python.csv", index=False)
print("\n✓ Results saved to 'mixed_effects_results_summary_python.csv'")

# ====================================================================
# 8. VISUALIZATIONS
# ====================================================================

print("\n" + "=" * 70)
print("CREATING VISUALIZATIONS")
print("=" * 70)

# 1. Mean trajectories by intervention group
plt.figure(figsize=(10, 6))
for status in df['intervention_status'].unique():
    subset = df[df['intervention_status'] == status]
    grouped = subset.groupby('time_years')['b1_overall_performance'].agg(['mean', 'sem'])
    plt.errorbar(grouped.index, grouped['mean'], yerr=grouped['sem'], 
                 marker='o', markersize=8, linewidth=2, capsize=5, 
                 label=status)

plt.xlabel('Time (Years)', fontsize=12)
plt.ylabel('Overall Performance Rating', fontsize=12)
plt.title('Student Performance Over Time by Intervention Status', fontsize=14, fontweight='bold')
plt.legend(title='Group', fontsize=11)
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig('plot1_trajectories_python.png', dpi=300, bbox_inches='tight')
print("✓ Saved plot1_trajectories_python.png")
plt.close()

# 2. Performance by region
plt.figure(figsize=(10, 6))
df_plot = df[['region', 'intervention_status', 'b1_overall_performance']].copy()
sns.boxplot(data=df_plot, x='region', y='b1_overall_performance', 
            hue='intervention_status', palette='Set2')
plt.xlabel('Region', fontsize=12)
plt.ylabel('Overall Performance Rating', fontsize=12)
plt.title('Student Performance by Region and Intervention Status', fontsize=14, fontweight='bold')
plt.legend(title='Group', fontsize=11)
plt.tight_layout()
plt.savefig('plot2_regional_python.png', dpi=300, bbox_inches='tight')
print("✓ Saved plot2_regional_python.png")
plt.close()

# 3. Predicted values plot (from best model)
plt.figure(figsize=(10, 6))
df_pred = df.copy()
df_pred['predicted'] = result3.fittedvalues

for status in df_pred['intervention_status'].unique():
    subset = df_pred[df_pred['intervention_status'] == status]
    grouped = subset.groupby('time_years')['predicted'].mean()
    plt.plot(grouped.index, grouped.values, marker='o', markersize=8, 
             linewidth=2, label=status)

plt.xlabel('Time (Years)', fontsize=12)
plt.ylabel('Predicted Performance Rating', fontsize=12)
plt.title('Predicted Performance: Intervention × Time Interaction', fontsize=14, fontweight='bold')
plt.legend(title='Group', fontsize=11)
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.savefig('plot3_interaction_python.png', dpi=300, bbox_inches='tight')
print("✓ Saved plot3_interaction_python.png")
plt.close()

# 4. All outcomes comparison
fig, axes = plt.subplots(2, 4, figsize=(16, 8))
axes = axes.flatten()

for i, outcome in enumerate(outcomes):
    df_plot = df[['intervention_status', outcome]].copy()
    df_plot = df_plot.dropna()
    
    sns.boxplot(data=df_plot, x='intervention_status', y=outcome, 
                palette='Set2', ax=axes[i])
    axes[i].set_title(outcome.replace('_', ' ').title(), fontsize=10)
    axes[i].set_xlabel('')
    axes[i].set_ylabel('Rating', fontsize=9)
    axes[i].tick_params(axis='x', rotation=45, labelsize=8)

plt.suptitle('All Performance Outcomes by Intervention Status', 
             fontsize=14, fontweight='bold', y=1.00)
plt.tight_layout()
plt.savefig('plot4_all_outcomes_python.png', dpi=300, bbox_inches='tight')
print("✓ Saved plot4_all_outcomes_python.png")
plt.close()

# 5. Heatmap of results
plt.figure(figsize=(10, 6))
heatmap_data = results_df[['Outcome', 'Intervention_Effect', 'Time_Effect', 'Interaction_Effect']].copy()
heatmap_data = heatmap_data.set_index('Outcome')
heatmap_data.index = [x.replace('_', ' ').replace('b1 ', '').replace('b2 ', '').replace('b3 ', '').replace('b4 ', '').replace('b5 ', '').replace('b6 ', '').replace('b7 ', '').replace('b8 ', '').title() for x in heatmap_data.index]

sns.heatmap(heatmap_data.T, annot=True, fmt='.3f', cmap='RdBu_r', center=0, 
            cbar_kws={'label': 'Effect Size'}, linewidths=0.5)
plt.title('Effect Sizes Across All Outcomes', fontsize=14, fontweight='bold')
plt.xlabel('')
plt.ylabel('Effect Type', fontsize=12)
plt.tight_layout()
plt.savefig('plot5_heatmap_python.png', dpi=300, bbox_inches='tight')
print("✓ Saved plot5_heatmap_python.png")
plt.close()

# ====================================================================
# 9. INTERPRETATION
# ====================================================================

print("\n" + "=" * 70)
print("INTERPRETATION GUIDE")
print("=" * 70)

params = result3.params
pvalues = result3.pvalues

print("\nKey Findings:")
print(f"1. Intervention Effect: {params.get('intervention', np.nan):.3f} "
      f"(p = {pvalues.get('intervention', np.nan):.4f})")
print(f"   → Students in intervention schools score {params.get('intervention', 0):.2f} "
      f"points higher on performance rating")

print(f"\n2. Time Effect: {params.get('time_years', np.nan):.3f} "
      f"(p = {pvalues.get('time_years', np.nan):.4f})")
print(f"   → Performance changes by {params.get('time_years', 0):.2f} points per year")

if 'intervention:time_years' in params.index:
    print(f"\n3. Intervention × Time Interaction: "
          f"{params.get('intervention:time_years', np.nan):.3f} "
          f"(p = {pvalues.get('intervention:time_years', np.nan):.4f})")
    print(f"   → The intervention effect changes by "
          f"{params.get('intervention:time_years', 0):.2f} points per year")
    
    if pvalues.get('intervention:time_years', 1) < 0.05:
        print("   → ✓ SIGNIFICANT: Intervention effectiveness changes over time!")

print("\n" + "=" * 70)
print("ANALYSIS COMPLETE")
print("=" * 70)
print("\nCheck your working directory for saved plots and results.")
print("\nFiles created:")
print("  - mixed_effects_results_summary_python.csv")
print("  - plot1_trajectories_python.png")
print("  - plot2_regional_python.png")
print("  - plot3_interaction_python.png")
print("  - plot4_all_outcomes_python.png")
print("  - plot5_heatmap_python.png")

File: srcs/data_plat_digiedu/mixed_effects_analysis.R
# ====================================================================
# MIXED EFFECTS MODEL ANALYSIS
# Teacher Perception of Student Performance Study
# ====================================================================

# Load required packages
# Install if needed: install.packages(c("lme4", "lmerTest", "ggplot2", "dplyr", "tidyr", "performance", "sjPlot"))

library(lme4)       # Mixed effects models
library(lmerTest)   # P-values for mixed models
library(ggplot2)    # Visualization
library(dplyr)      # Data manipulation
library(tidyr)      # Data reshaping
library(performance) # Model diagnostics and R-squared
library(sjPlot)     # Nice model tables and plots

# ====================================================================
# 1. LOAD AND PREPARE DATA
# ====================================================================

# Load data
df <- read.csv("data_structure_template.csv", stringsAsFactors = FALSE)

# Convert variables to appropriate types
df <- df %>%
  mutate(
    # Convert to factors
    teacher_id = as.factor(teacher_id),
    school_id = as.factor(school_id),
    region = as.factor(region),
    intervention_status = as.factor(intervention_status),
    teaching_experience_years = as.numeric(teaching_experience_years),
    class_size = as.numeric(class_size),
    
    # Convert date
    survey_date = as.Date(survey_date),
    
    # Create time variable (years from first survey)
    time_years = as.numeric(difftime(survey_date, min(survey_date), units = "days")) / 365.25,
    
    # Recode intervention status (Treatment=1, Control=0)
    intervention = ifelse(intervention_status == "Treatment", 1, 0)
  )

# Check data structure
str(df)
summary(df)

# Check for missing data
sum(is.na(df))
colSums(is.na(df))

# ====================================================================
# 2. DESCRIPTIVE STATISTICS
# ====================================================================

# Sample sizes
cat("\n=== SAMPLE SIZES ===\n")
cat("Total responses:", nrow(df), "\n")
cat("Number of teachers:", length(unique(df$teacher_id)), "\n")
cat("Number of schools:", length(unique(df$school_id)), "\n")
cat("Number of regions:", length(unique(df$region)), "\n")

# By intervention status
table(df$intervention_status)
table(df$region, df$intervention_status)

# Means by group
cat("\n=== DESCRIPTIVE STATISTICS BY INTERVENTION STATUS ===\n")
df %>%
  group_by(intervention_status) %>%
  summarise(
    n = n(),
    mean_b1 = mean(b1_overall_performance, na.rm = TRUE),
    sd_b1 = sd(b1_overall_performance, na.rm = TRUE),
    mean_b2 = mean(b2_problem_solving, na.rm = TRUE),
    sd_b2 = sd(b2_problem_solving, na.rm = TRUE),
    mean_b6 = mean(b6_engagement, na.rm = TRUE),
    sd_b6 = sd(b6_engagement, na.rm = TRUE)
  )

# By region
cat("\n=== DESCRIPTIVE STATISTICS BY REGION ===\n")
df %>%
  group_by(region) %>%
  summarise(
    n = n(),
    mean_b1 = mean(b1_overall_performance, na.rm = TRUE),
    sd_b1 = sd(b1_overall_performance, na.rm = TRUE)
  )

# ====================================================================
# 3. NULL MODEL (Unconditional Means Model)
# ====================================================================

cat("\n\n=== NULL MODEL: Partitioning Variance ===\n")

# Outcome: Overall Performance (B1)
model0_b1 <- lmer(b1_overall_performance ~ 1 + (1|region/school_id), 
                  data = df, REML = TRUE)
summary(model0_b1)

# Calculate ICC (Intraclass Correlation Coefficient)
variance_components <- as.data.frame(VarCorr(model0_b1))
var_region <- variance_components$vcov[1]
var_school <- variance_components$vcov[2]
var_residual <- variance_components$vcov[3]

ICC_region <- var_region / (var_region + var_school + var_residual)
ICC_school <- var_school / (var_region + var_school + var_residual)
ICC_total <- (var_region + var_school) / (var_region + var_school + var_residual)

cat("\n--- Variance Components ---\n")
cat("Region variance:", round(var_region, 3), "\n")
cat("School variance:", round(var_school, 3), "\n")
cat("Residual variance:", round(var_residual, 3), "\n")
cat("\n--- Intraclass Correlations ---\n")
cat("ICC (Region):", round(ICC_region, 3), "- proportion of variance due to regions\n")
cat("ICC (School):", round(ICC_school, 3), "- proportion of variance due to schools\n")
cat("ICC (Total):", round(ICC_total, 3), "- proportion of variance due to clustering\n")

# Interpretation
if (ICC_total > 0.05) {
  cat("\n✓ ICC > 0.05: Mixed effects model is justified!\n")
} else {
  cat("\n⚠ ICC < 0.05: Consider whether mixed effects model is necessary\n")
}

# ====================================================================
# 4. MODEL 1: INTERVENTION AND TIME EFFECTS
# ====================================================================

cat("\n\n=== MODEL 1: Main Effects of Intervention and Time ===\n")

model1_b1 <- lmer(b1_overall_performance ~ intervention + time_years + 
                    (1|region/school_id), 
                  data = df, REML = FALSE)
summary(model1_b1)

# Model fit
performance::r2(model1_b1)

# ====================================================================
# 5. MODEL 2: ADD INTERVENTION × TIME INTERACTION
# ====================================================================

cat("\n\n=== MODEL 2: Intervention × Time Interaction ===\n")

model2_b1 <- lmer(b1_overall_performance ~ intervention * time_years + 
                    (1|region/school_id), 
                  data = df, REML = FALSE)
summary(model2_b1)

# Model fit
performance::r2(model2_b1)

# Test if interaction improves model
anova(model1_b1, model2_b1)

# ====================================================================
# 6. MODEL 3: ADD REGIONAL DIFFERENCES
# ====================================================================

cat("\n\n=== MODEL 3: Add Region as Fixed Effect ===\n")

model3_b1 <- lmer(b1_overall_performance ~ intervention * time_years + region + 
                    (1|school_id), 
                  data = df, REML = FALSE)
summary(model3_b1)

# Model comparison
anova(model2_b1, model3_b1)

# ====================================================================
# 7. MODEL 4: ADD COVARIATES (FULL MODEL)
# ====================================================================

cat("\n\n=== MODEL 4: Full Model with Covariates ===\n")

# Center continuous predictors for easier interpretation
df <- df %>%
  mutate(
    teaching_exp_centered = teaching_experience_years - mean(teaching_experience_years, na.rm = TRUE),
    class_size_centered = class_size - mean(class_size, na.rm = TRUE)
  )

model4_b1 <- lmer(b1_overall_performance ~ 
                    intervention * time_years + 
                    region + 
                    teaching_exp_centered + 
                    class_size_centered +
                    (1|school_id), 
                  data = df, REML = FALSE)
summary(model4_b1)

# Model fit
performance::r2(model4_b1)

# Model comparison
anova(model3_b1, model4_b1)

# ====================================================================
# 8. MODEL 5: TEST REGION × INTERVENTION INTERACTION
# ====================================================================

cat("\n\n=== MODEL 5: Region × Intervention Interaction ===\n")

model5_b1 <- lmer(b1_overall_performance ~ 
                    intervention * time_years + 
                    intervention * region +
                    teaching_exp_centered + 
                    class_size_centered +
                    (1|school_id), 
                  data = df, REML = FALSE)
summary(model5_b1)

# Test if interaction is significant
anova(model4_b1, model5_b1)

# ====================================================================
# 9. MODEL DIAGNOSTICS
# ====================================================================

cat("\n\n=== MODEL DIAGNOSTICS ===\n")

# Choose best model (e.g., model4_b1)
best_model <- model4_b1

# Residual plots
par(mfrow=c(2,2))

# 1. Residuals vs Fitted
plot(fitted(best_model), residuals(best_model),
     xlab="Fitted Values", ylab="Residuals",
     main="Residuals vs Fitted")
abline(h=0, col="red", lty=2)

# 2. Q-Q plot of residuals
qqnorm(residuals(best_model))
qqline(residuals(best_model), col="red")

# 3. Scale-Location plot
plot(fitted(best_model), sqrt(abs(residuals(best_model))),
     xlab="Fitted Values", ylab="√|Residuals|",
     main="Scale-Location")

# 4. Q-Q plot of random effects (school)
school_ranef <- ranef(best_model)$school_id[,1]
qqnorm(school_ranef, main="Q-Q Plot: School Random Effects")
qqline(school_ranef, col="red")

par(mfrow=c(1,1))

# Check for influential observations
# Influence diagnostics (if needed)
# library(influence.ME)
# influence_data <- influence(best_model, obs = TRUE)

# ====================================================================
# 10. ANALYZE ALL OUTCOME VARIABLES
# ====================================================================

cat("\n\n=== MODELS FOR ALL OUTCOMES ===\n")

# List of outcome variables
outcomes <- c("b1_overall_performance", "b2_problem_solving", 
              "b3_critical_thinking", "b4_collaboration",
              "b5_communication", "b6_engagement", 
              "b7_behavior", "b8_persistence")

# Store results
results_list <- list()

for (outcome in outcomes) {
  cat("\n--- Outcome:", outcome, "---\n")
  
  # Create formula
  formula_str <- paste(outcome, "~ intervention * time_years + region + 
                       teaching_exp_centered + class_size_centered + (1|school_id)")
  
  # Fit model
  model <- lmer(as.formula(formula_str), data = df, REML = FALSE)
  
  # Store results
  results_list[[outcome]] <- model
  
  # Print summary
  print(summary(model))
  print(performance::r2(model))
}

# ====================================================================
# 11. CREATE RESULTS TABLE
# ====================================================================

# Extract key coefficients from all models
results_table <- data.frame(
  Outcome = outcomes,
  Intercept = numeric(length(outcomes)),
  Intervention_Effect = numeric(length(outcomes)),
  Intervention_p = numeric(length(outcomes)),
  Time_Effect = numeric(length(outcomes)),
  Time_p = numeric(length(outcomes)),
  Interaction_Effect = numeric(length(outcomes)),
  Interaction_p = numeric(length(outcomes)),
  Marginal_R2 = numeric(length(outcomes)),
  Conditional_R2 = numeric(length(outcomes))
)

for (i in seq_along(outcomes)) {
  model <- results_list[[outcomes[i]]]
  coefs <- summary(model)$coefficients
  r2_values <- performance::r2(model)
  
  results_table$Intercept[i] <- round(coefs[1, 1], 3)
  results_table$Intervention_Effect[i] <- round(coefs[2, 1], 3)
  results_table$Intervention_p[i] <- round(coefs[2, 5], 4)
  results_table$Time_Effect[i] <- round(coefs[3, 1], 3)
  results_table$Time_p[i] <- round(coefs[3, 5], 4)
  
  # Check if interaction exists
  if (nrow(coefs) >= 4) {
    results_table$Interaction_Effect[i] <- round(coefs[4, 1], 3)
    results_table$Interaction_p[i] <- round(coefs[4, 5], 4)
  }
  
  results_table$Marginal_R2[i] <- round(r2_values$R2_marginal, 3)
  results_table$Conditional_R2[i] <- round(r2_values$R2_conditional, 3)
}

cat("\n\n=== SUMMARY RESULTS TABLE ===\n")
print(results_table)

# Save results
write.csv(results_table, "mixed_effects_results_summary.csv", row.names = FALSE)

# ====================================================================
# 12. VISUALIZATIONS
# ====================================================================

cat("\n\n=== CREATING VISUALIZATIONS ===\n")

# 1. Mean trajectories by intervention group
plot1 <- ggplot(df, aes(x = time_years, y = b1_overall_performance, 
                        color = intervention_status, group = intervention_status)) +
  stat_summary(fun = mean, geom = "line", size = 1.2) +
  stat_summary(fun = mean, geom = "point", size = 3) +
  stat_summary(fun.data = mean_se, geom = "errorbar", width = 0.1) +
  labs(title = "Student Performance Over Time by Intervention Status",
       x = "Time (Years)", 
       y = "Overall Performance Rating",
       color = "Group") +
  theme_bw() +
  theme(text = element_text(size = 12))

print(plot1)
ggsave("plot1_trajectories.png", plot1, width = 8, height = 6)

# 2. Performance by region
plot2 <- ggplot(df, aes(x = region, y = b1_overall_performance, fill = intervention_status)) +
  geom_boxplot(alpha = 0.7) +
  labs(title = "Student Performance by Region and Intervention Status",
       x = "Region", 
       y = "Overall Performance Rating",
       fill = "Group") +
  theme_bw() +
  theme(text = element_text(size = 12))

print(plot2)
ggsave("plot2_regional.png", plot2, width = 8, height = 6)

# 3. Interaction plot
# Predicted values
df$predicted_b1 <- predict(model4_b1)

plot3 <- ggplot(df, aes(x = time_years, y = predicted_b1, 
                        color = intervention_status, group = intervention_status)) +
  geom_line(size = 1.2) +
  labs(title = "Predicted Performance: Intervention × Time Interaction",
       subtitle = "Based on Mixed Effects Model",
       x = "Time (Years)", 
       y = "Predicted Performance Rating",
       color = "Group") +
  theme_bw() +
  theme(text = element_text(size = 12))

print(plot3)
ggsave("plot3_interaction.png", plot3, width = 8, height = 6)

# 4. Caterpillar plot of school random effects
ranef_df <- as.data.frame(ranef(model4_b1)$school_id)
ranef_df$school_id <- rownames(ranef_df)
ranef_df <- ranef_df %>% arrange(`(Intercept)`)
ranef_df$school_id <- factor(ranef_df$school_id, levels = ranef_df$school_id)

plot4 <- ggplot(ranef_df, aes(x = `(Intercept)`, y = school_id)) +
  geom_point(size = 2) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "red") +
  labs(title = "School-Level Random Effects (Caterpillar Plot)",
       subtitle = "Deviation from overall mean",
       x = "Random Effect (Intercept)", 
       y = "School ID") +
  theme_bw() +
  theme(text = element_text(size = 10))

print(plot4)
ggsave("plot4_caterpillar.png", plot4, width = 8, height = 10)

# 5. All outcomes comparison
outcomes_long <- df %>%
  select(teacher_id, intervention_status, time_years, all_of(outcomes)) %>%
  pivot_longer(cols = all_of(outcomes), names_to = "Outcome", values_to = "Rating")

plot5 <- ggplot(outcomes_long, aes(x = intervention_status, y = Rating, fill = intervention_status)) +
  geom_boxplot(alpha = 0.7) +
  facet_wrap(~ Outcome, scales = "free_y") +
  labs(title = "All Performance Outcomes by Intervention Status",
       x = "Intervention Status", 
       y = "Rating",
       fill = "Group") +
  theme_bw() +
  theme(text = element_text(size = 10),
        axis.text.x = element_text(angle = 45, hjust = 1))

print(plot5)
ggsave("plot5_all_outcomes.png", plot5, width = 12, height = 8)

# ====================================================================
# 13. PUBLICATION-READY TABLE
# ====================================================================

# Use sjPlot for nice HTML table
sjPlot::tab_model(model4_b1, 
                  show.re.var = TRUE,
                  show.icc = TRUE,
                  show.r2 = TRUE,
                  file = "model_table.html")

cat("\n✓ HTML table saved as 'model_table.html'\n")

# ====================================================================
# 14. INTERPRETATION HELPER
# ====================================================================

cat("\n\n=== INTERPRETATION GUIDE ===\n")

coefs <- summary(model4_b1)$coefficients

cat("\nKey Findings:\n")
cat("1. Intervention Effect:", round(coefs["intervention", "Estimate"], 3), 
    " (p =", round(coefs["intervention", "Pr(>|t|)"], 4), ")\n")
cat("   → Students in intervention schools score", round(coefs["intervention", "Estimate"], 2), 
    "points higher on performance rating\n")

cat("\n2. Time Effect:", round(coefs["time_years", "Estimate"], 3), 
    " (p =", round(coefs["time_years", "Pr(>|t|)"], 4), ")\n")
cat("   → Performance changes by", round(coefs["time_years", "Estimate"], 2), 
    "points per year\n")

if ("intervention:time_years" %in% rownames(coefs)) {
  cat("\n3. Intervention × Time Interaction:", 
      round(coefs["intervention:time_years", "Estimate"], 3), 
      " (p =", round(coefs["intervention:time_years", "Pr(>|t|)"], 4), ")\n")
  cat("   → The intervention effect changes by", 
      round(coefs["intervention:time_years", "Estimate"], 2), 
      "points per year\n")
  
  if (coefs["intervention:time_years", "Pr(>|t|)"] < 0.05) {
    cat("   → ✓ SIGNIFICANT: Intervention effectiveness changes over time!\n")
  }
}

r2_vals <- performance::r2(model4_b1)
cat("\n4. Model Fit:\n")
cat("   Marginal R² =", round(r2_vals$R2_marginal, 3), 
    "(variance explained by fixed effects)\n")
cat("   Conditional R² =", round(r2_vals$R2_conditional, 3), 
    "(variance explained by entire model)\n")

cat("\n\n=== ANALYSIS COMPLETE ===\n")
cat("Check your working directory for saved plots and tables.\n")






